src/socket.c

```diff
@@ -22,112 +24,159 @@
    \return The new socket's file descriptor
 */
 static int bind_socket(const fastd_bind_address_t *addr) {
+	const int zero = 0;
+	const int one = 1;
 	int fd = -1;
 	int af = AF_UNSPEC;
+	fastd_peer_address_t bind_address = addr->addr;
 
-	if (addr->addr.sa.sa_family != AF_INET) {
+	const int v6only = bind_address.sa.sa_family == AF_INET6 || addr->sourceaddr.sa.sa_family == AF_INET6;
+	if (!fastd_peer_address_is_v4_multicast(&bind_address)) {
 		fd = socket(PF_INET6, SOCK_DGRAM | SOCK_NONBLOCK, IPPROTO_UDP);
 		if (fd >= 0) {
 			af = AF_INET6;
 
-			int val = (addr->addr.sa.sa_family == AF_INET6);
-			if (setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, &val, sizeof(val))) {
-				pr_warn_errno("setsockopt");
+			if (setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, &v6only, sizeof(v6only))) {
+				pr_warn_errno("setsockopt: unable to set socket for IPv6 only");
 				goto error;
 			}
 		}
 	}
-	if (fd < 0 && addr->addr.sa.sa_family != AF_INET6) {
+
+	if (fd < 0 && bind_address.sa.sa_family != AF_INET6 && addr->sourceaddr.sa.sa_family != AF_INET6) {
 		fd = socket(PF_INET, SOCK_DGRAM | SOCK_NONBLOCK, IPPROTO_UDP);
-		if (fd < 0)
-			exit_errno("unable to create socket");
-		else
+		if (fd >= 0)
 			af = AF_INET;
 	}
 
-	if (fd < 0)
+	if (fd < 0) {
+		pr_error_errno("socket: unable to create socket");
 		goto error;
+	}
+
+	if ((bind_address.sa.sa_family == AF_UNSPEC && (af == AF_INET || addr->sourceaddr.sa.sa_family == AF_INET)) ||
+			fastd_peer_address_is_v4_multicast(&bind_address)) {
+		bind_address.in.sin_family = AF_INET;
+		bind_address.in.sin_addr.s_addr = INADDR_ANY;
+
+		if (af == AF_INET6)
+			fastd_peer_address_widen(&bind_address);
+	} else if (bind_address.sa.sa_family == AF_UNSPEC || fastd_peer_address_is_v6_multicast(&bind_address)) {
+		bind_address.in6.sin6_family = AF_INET6;
+		bind_address.in6.sin6_addr = in6addr_any;
+
+		if (addr->addr.sa.sa_family == AF_UNSPEC)
+			bind_address.in6.sin6_port = addr->addr.in.sin_port;
+	} else if (af == AF_INET6)
+		fastd_peer_address_widen(&bind_address);
 
 #ifdef NO_HAVE_SOCK_NONBLOCK
 	fastd_setnonblock(fd);
 #endif
 
-	int one = 1;
-
 #ifdef USE_PKTINFO
-	if (setsockopt(fd, IPPROTO_IP, IP_PKTINFO, &one, sizeof(one))) {
+	if ((af == AF_INET || !v6only) && setsockopt(fd, IPPROTO_IP, IP_PKTINFO, &one, sizeof(one))) {
 		pr_error_errno("setsockopt: unable to set IP_PKTINFO");
 		goto error;
 	}
 #endif
 
+	if (af != AF_INET && setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &one, sizeof(one))) {
+		pr_error_errno("setsockopt: unable to set IPV6_RECVPKTINFO");
+		goto error;
+	}
+
 #ifdef USE_FREEBIND
+#ifdef USE_FREEBIND6
+	if ((af == AF_INET || !v6only) && setsockopt(fd, IPPROTO_IP, IP_FREEBIND, &one, sizeof(one)))
+		pr_warn_errno("setsockopt: unable to set IP_FREEBIND");
+	if (af != AF_INET && setsockopt(fd, IPPROTO_IPV6, IPV6_FREEBIND, &one, sizeof(one)))
+		pr_warn_errno("setsockopt: unable to set IPV6_FREEBIND");
+#else
```

Comment:
What's the advantage of `IPV6_FREEBIND`? I can't find it in the documentation (`man 7 ipv6` as of Linux man-pages 5.10), so I think it should not be used unless necessary.